core/src/main/java/com/linecorp/armeria/internal/common/stream/DecodedHttpStreamMessage.java

```diff
@@ -157,14 +157,22 @@ protected void onRequest(long n) {
         }
     }
 
-    private void askUpstreamForElement() {
+    /**
+     * Directly requests n elements to the {@code streamMessage} that was specified when creating this
+     * {@link StreamMessage}. It is useful when the {@link HttpDecoder} emits another {@link StreamMessage}s.
+     */
+    public void askUpstreamForElements(long n) {
         if (!askedUpstreamForElement) {
```

Comment:
A request for elements might be lost if `askedUpstreamForElement` is true.
How about separately tracking the directly requested number of elements?

For example:
```java
// Should be updated in an event loop
private long directRequest;

public void askUpstreamForElements(long n) {
    // Make sure that this API should be called after subscribing.
    assert executor != null;
    // Update in an event loop which subscribed this stream message.
    directRequest += n;
    // Call the original API.
    askUpstreamForElement();
}

private final class HttpMessageSubscriber implements Subscriber<HttpObject> {

    public void onNext(HttpObject obj) {
        ...
        if (directRequest > 0) {
            directRequest--;
        }
        // If there is a remaining demand, ask upstream for more elements
        if (directRequest > 0) {
            askUpstreamForElement();
            return;
        }
        
        
        if (handlerProduced) {
             // Keep the original code path.
             ...
        }
    }
```

I'm not 100% sure the example code correctly works. But it might be worth to try it. 😄